Initial monero seeds/mnemonics and address integration

Bowler-Bear opened the pull request:
This resolves issues #4 , #5 , and #17 .

Monero Mnemonics
The monero mnemonics generated are the same as the mnemonics generated by the official cli and gui wallet. These mnemonics are referenced as 'legacy' seeds in the official documentation, so the underlying c library was named legacy_monero_mnemonic. A moneromnemonics micropython module was introduced to contain seeds functionality related to monero. A 'legacy' submodule to moneromnemonics was also introduced that allows access to the legacy_monero_mnemonic functions through the python code. Previous calls to bitcoin related mnemonics in the flows and tasks were replaced with calls to the legacy module. The restore flow was allowed to have 24 words since they can technically be valid as long as the words are valid seed words(the 25th word is a checksum), but the code will still work the same if the 24 word option is removed.

Word Prediction
To reduce major refactors to the code, word prediction was incorporated into the existing foundation.bip39 micropython module. The bip39_word_info.c word list had to have most entries removed to offset the increase in firmware size from the addition of the monero_english_word_info word list.

Stashed Values
The seed is stored, but monero does not use the same hierarchical key structure as bitcoin. Monero uses a major/minor index pair for subaddress generation, so the stored xpriv and master secret were no longer needed and removed from the encode/decode functionality. However, the code is heavily intertwined with the xfingerprint and xpub settings references and attempting to remove them from the SensitiveValues class caused many errors, so they were left in place for now as their existence don't adversely effect the functionality of the introduced monero code.

View Wallet Export
The existing wallet exports were removed as they do not support monero and increase the overall firmware binary size. A Monero CLI export entry was added in their place. It exports the private view key and public address in json format. These values can be used with both the official cli and gui wallets to create a view-only hot wallet based on the passport's cold wallet.

Address Verification
Monero addreesses can now be verified as belonging to this seed/wallet, Primary and integrated addresses only use index (0, 0) so they are only checked against that index. Subaddresses are checked along the minor index, as iterator i, in groups of 100 starting at 0 with the current visible account and it's number acting as the major index. They follow the form (account number, i). For account 0, the searching starts at minor index 1 because (0, 0) is always a primary index.

jonathancross commented:
Thanks @Bowler-Bear  -- thrilling to see progress on this!

One thought: It might be easier for @mjg-foundation to review if broken into smaller pieces?

mjg-foundation commented:
> Thanks @Bowler-Bear -- thrilling to see progress on this!
> 
> One thought: It might be easier for @mjg-foundation to review if broken into smaller pieces?

No need to break it up just yet, somehow I missed any notifications of this PR until your comment here.

@Bowler-Bear have you run this on the simulator to test it? Could you write up a quick testing guide so i can validate things? I'm guessing you used anonero or the monero CLI to validate the exports and generate addresses to verify. Note that the bounty on #17 requires exporting via UR and microsd, but this should be easy now that you've got the CLI export format working.

Bowler-Bear commented:
I have tested it through the simulator. I'll write up a some steps that can be followed to validate that it is working, at least through the simulator. I don't have the physical device, so I couldn't test it on there. Although, aside from flashing the firmware, the testing steps should be the same between the two. I initially used the monero CLI to generate and validate mnemonics and seed values. Later when working on the exports and address verification I switched to the monero GUI because it has built in qr codes that are generated for it's addresses.

I added another commit that allows the export format to be a UR code, but it's an encoded json string that contains the keys needed to generate a view-only wallet. However, neither the CLI nor the GUI directly support UR codes. They both prompt the user for their private view key and public address to generate the wallet. So anyone using that UR export would have to decode it in a separate program, then use the decoded string the same way the json file exported to the microsd would be used. I might need more clarification on what you want for the UR export if using a json string isn't sufficient.

I also added another commit that removes an error when restoring using a backup file. Issue 4 doesn't really mention backup support so I had missed it earlier, but I ran into it while writing out the testing steps that included backup files. I figured I should fix it in case anyone wants to use backup files with this software as well.

Bowler-Bear commented:
Apparently, the exported key names I used were slightly off from those expected from Anonero. I've adjusted them and added Anonero as it's own export.

mjg-foundation commented:
> Apparently, the exported key names I used were slightly off from those expected from Anonero. I've adjusted them and added Anonero as it's own export.

Cool! I'll get to testing and reviewing this over the next week, I'm currently kept pretty busy by official business at foundation, but I'll make sure to pay out if everything goes well

Bowler-Bear commented:
> @Bowler-Bear Please post an XMR address somewhere just in case some unknown stranger wants to tip you for your fabulous work here.

Anything related to this pull request can be sent to 8A4enbC9VrxaYCQ7uQGSKJ8Us3H7CauUHaewQ8ZYNstmjP9RzxKoR25KajkzHV3Hr4aMiiyC7Nm5DPDCgaDx1U8jQp3xHQz. Any tips by anyone coming across this pull would be unexpected but greatly appreciated.

Bowler-Bear commented:
@mjg-foundation Any updates on your review?

Bowler-Bear commented:
I was using `just build color` in the `ports/stm32` folder as described in DEVELOPMENT.md, but I was already working in an Ubuntu 20.04 environment. I figured it would build fine in docker as well. When I get a chance a little later, I'll build it through docker too and see if I run into any issues.

Bowler-Bear commented:
@mjg-foundation So last night I went ahead and built it through docker. It was initially unable to build `cbindgen` and `just` while setting up the docker image. The problem was that it was trying to build more recent dependencies that required a newer rust toolchain than 1.67.1. I had to change the Dockerfile commands for the cargo installs to `cargo install --locked cbindgen@0.24.1` and `cargo install --locked just@1.13.0`. Aside from that there were no other issues I encountered. I was able to build the firmware using `just build-firmware color`. If these weren't the issues you were facing, could you tell me more about what issues you're facing?

Bowler-Bear commented:
@mjg-foundation were you able to build the firmware?

mjg-foundation commented:
> @mjg-foundation were you able to build the firmware?

Could you push your changes to the Dockerfile?

mjg-foundation commented:
Got it building and installed it, got stuck in a boot loop with this message:
```
================================================================================
Start main.py:
Available:     246,384 of 264,960 bytes
Largest Block: 243,168 bytes
================================================================================
Traceback (most recent call last):
  File "main.py", line 175, in <module>
  File "main.py", line 25, in go
  File "tasks/__init__.py", line 31, in <module>
  File "tasks/get_seed_words_task.py", line 8, in <module>
ImportError: no module named 'monero_mnemonic_languages'
MPY: soft reboot
[Passport_board_init]
480000000, 480000000, 240000000, 120000000, 120000000
MPY: can't mount SD card
```
Fixed it by adding monero_mnemonic_languages.py to modules/manifest.py. I was able to recover my dev board using the STlink, so no big deal.

@Bowler-Bear make sure to push the change and have me verify the new firmware before loading it on a personal device, in case the fix isn't right.

So far I've been able to load a known seed and confirm that address verification and account exports work, knocking out #4 #5 and #17 . I'll also test making a random seed, but it's looking like everyone in on these bounties should prepare their payments. I'll see if I can upload a firmware that people can try out safely on their devices.

@donttracemebruh @vdo

mjg-foundation commented:
I've made this [developer pubkey](https://drive.google.com/file/d/17hHeoJDIV_Kv6_EiCmFrbJusleMbYCrK/view?usp=sharing), and using [these instructions](https://docs.foundation.xyz/passport/passport-menu/settings/advanced/#developer-pubkey), installed this [monero firmware](https://drive.google.com/file/d/1jFVJpBIWtK0t66F0lGLr6QNg5IkkcJqQ/view?usp=sharing), and successfully went back to official bitcoin firmware afterwards.

If you do this on your device, make **absolutely sure** your bitcoin seed is backed up, since I haven't thoroughly tested if seeds survive switching back and forth between these firmwares.

I just saw @Bowler-Bear pushed new changes, these don't include those, I'll verify that the firmware code in those changes works shortly.

Bowler-Bear commented:
Good catch on the unfrozen file. I believe the simulator uses an edited environment path to load modules, bypassing the manifest, so that didn't show up as an error for me. I've pushed the changes. Let me know if there is anything else that needss to be addressed.

mjg-foundation commented:
I wasn't able to build the dockerfile you pushed, I think the '^' symbol needs to be taken out of the cargo install lines, it builds when I do that. You could also just bump toolchain up to 1.70, which is what the official firmware [Dockerfile](https://github.com/Foundation-Devices/passport2/blob/main/Dockerfile) is using now, but we still lock `just` without a '^" indicating minor updates allowed.

I'll probably make a separate issue for it, but there are a bunch of references to bitcoin throughout the UI that could be cleaned up. One is in onboarding, which no one will be going through with a monero firmware installed, and the others are in the Monero menu, where the unit type options are "Sats" and "Bitcoin", but these are tied into the whole chaintype system, so it may be nontrivial to change this.

mjg-foundation commented:
@Bowler-Bear  I'm not sure why, but the change you pushed to fix boot loop aren't working for me. Maybe the module needs to be included in a certain order. I fixed it originally by including it at the start of the freeze() block titled Xmr. Moving it from where you placed it to this position worked again to fix the boot loop, so let's go with that:
```
# Xmr
freeze('$(MPY_DIR)/ports/stm32/boards/Passport/modules',
       ('monero_mnemonic_languages.py',
        'xmr/addresses.py',
        'xmr/crypto/__init__.py',
        'xmr/monero.py',
        'xmr/networks.py'))
```

mjg-foundation commented:
Assuming those changes go well, I'll do final testing to make sure random seeds can connect with the gui wallet and verify addresses as well, and also see if bitcoin seeds are preserved when jumping between firmware. This isn't a requirement for the bounties but definitely a quality of life thing I'm hoping just works.

Bowler-Bear commented:
You're right the carets are a problem. I had [previously removed them too](https://github.com/mjg-foundation/passport2-monero/pull/18#issuecomment-3041490613) but forgot to remove them again. I'll leave it up to you to bump up the toolchain. I'm not sure that belongs in this PR.

As far as the bitcoin code, some of it has to be gutted out anyway for signing to be implemented because the FLASH_TEXT memory region is already nearly full after this change. I had to purposely disable some bitcoin related modules to get it under size. I was planning on removing/refactoring it(some of the code is useful) as much of it as possible before working on the signing. Also, yes some of the code is tied together quite a bit and is not that simple to remove.

Swapping the firmwares to allow use of both BTC and XMR at the same time probably won't work as is. I made no changes to the slots used to save the seed bits to. I couldn't test for it through the simulator and I don't have the corresponding docs/information for the memory regions/slots of the Secure Element either. However, it should be possible to save the XMR seed bits to one the unused slots. I remember seeing somewhere in the codebase that there were several slots that were unused. IIRC when I looked at the code one was was an undocumented slot but the rest were unused. So one of those should be able to be used, allowing swapping OFW and dev FW without deleting the key data in between swaps. I can see how big of a QOL improvement that would be so I would be willing to help try getting it working as part of a separate pull, but you'll have to do the testing. I also can't guarantee I can get it working, but in theory it should be possible.

Bowler-Bear commented:
Got it. Yeah, that makes sense. In that case, I'll need tomake a change to stash.py, but that's doable.

mjg-foundation commented:
> I've tested the core functionality out for #4 #5 and #17 . I believe I owe @Bowler-Bear 2 XMR for #4 , 1 XMR for #5 , and 2 XMR for #17 . I've sent my payment to the address you posted above. I'll leave it to you to do the honors of pushing the merge button, I'm pretty sure it should be open to you since I've approved the PR.

@donttracemebruh owes 2 XMR for #4, and .5 XMR for #5 
@vdo owes 1 XMR for #4
